### PR TITLE
[wasm/wasi] Consolidate build targets

### PR DESCRIPTION
**Pull request by radical**
[wasm/wasi] Consolidate build targets

Current state of build files:
```
    wasm: WasmApp.props, WasmApp.targets, WasmApp.Native.targets
    wasi: WasiApp.props, WasiApp.targets, WasiApp.Native.targets
```

The wasm, and wasi build have lot of shared code but that is not
representative in the actual files, since the wasi targets started life
as a quick-copy-comment-out-bits of the wasm targets.

This commit consolidates these into:
```
    common: WasmApp.Common.props, WasmApp.Common.targets
    wasm : WasmApp.props, WasmApp.targets
    wasi : WasiApp.props, WasiApp.targets
```

 ## `WasmApp.Common.{props,targets}`

This has all the common parts of the build for browser-wasm, and wasi,
and includes bits from `WasmApp.{props,targets}`, and
`WasmApp.Native.{props,targets}`.

- The top level target remains the same - `WasmBuildApp`.
- There are a few "public" targets that can be hooked into:
    - `PrepareInputsForWasmBuild`
    - `WasmGenerateAppBundle`
    - `PrepareForWasmBuildNative`
    - `WasmLinkDotNet`

- all these public targets have corresponding `*DependsOn` properties
  which can be used for extending the build

note: this commit does not add a public target for AOT, but it might be
added in future.

 ## `WasmApp.{props,targets}`

This is for `browser-wasm` projects. The file might be renamed in
future.

 ## `WasiApp.{props,targets}`

This is for `wasi-wasm` projects. `ILStrip` becomes usable as a feature
for `wasi-wasm` because of this consolidation.

### CODE REVIEW

**Comment by radical** [comment]
https://github.com/dotnet/emsdk/pull/621 is needed for this.

**Comment by radical** [comment]
This is admittedly kinda hard to review, so I'm depending on the tests quite a bit. This does not change core specifics of the build, but moves around things, different public targets, and dependencies.

**Comment by radical** [comment]
TODO for follow up PRs:
- [ ] add wasi-sdk version check
- [ ]

File: src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+using Xunit.Abstractions;
+using Wasm.Build.Tests;
+
+#nullable enable
+
+namespace Wasi.Build.Tests;
+
+public class ILStripTests : BuildTestBase
+{
+    public ILStripTests(ITestOutputHelper output, SharedBuildPerTestClassFixture buildContext)
```
Comment (radical) [inline src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs:15]: cc @fanyang-mono

File: eng/testing/tests.wasm.targets [hunk 2]
```diff
@@ -109,6 +109,8 @@
       <BundleFiles Include="$(RuntimeConfigFilePath)"           TargetDir="publish" />
 
       <BundleFiles Include="$(WasmSharedPath)data\aot-tests\*"  TargetDir="publish" />
+      <!-- FIXME: what would be the correct place to do this? -->
```
Comment (pavelsavara) [inline eng/testing/tests.wasm.targets:112]: should we rename `wasm` -> `browser`  (OS) and use `wasm` for the (architecture) ?

Comment (radical) [reply eng/testing/tests.wasm.targets:112]: Yeah, I avoided those renames in this PR. We are following the pattern you suggested in most places now, like `tests.{wasm,browser,wasi}.targets`, and recently `sendtohelix-{wasm,browser,wasi}.proj`.

**Comment by radical** [comment]
For review, for the wasm case I think we have pretty good test coverage to be reasonably confident. There will be likely be some cases missed, or small regressions, which we can be fixed in follow up PRs.
For wasi, we have limited build support right now anyway, and *some* of that is covered by tests, and this PR makes it possible to have the same kinda features as the wasm build - for example, ilstrip, singlefilebundle (tbd).

**Review by ilonatommy — APPROVED** [review]

**Comment by vargaz** [comment]
Would be nice to review/merge this quickly so it doesn't conflict with other PRs, i.e.
https://github.com/dotnet/runtime/pull/95834

File: src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            testOutput.WriteLine ($"compressedStrippedAssembly_fi: {compressedStrippedAssembly_fi.Length}, {compressedStrippedAssembly}");
+
+            // original > stripped assembly
+            Assert.True(compressedOriginalAssembly_fi.Length > compressedStrippedAssembly_fi.Length,
+                        $"Expected original assembly ({compressedOriginalAssembly}) size ({compressedOriginalAssembly_fi.Length}) " +
+                        $"to be bigger than the stripped assembly ({compressedStrippedAssembly}) size ({compressedStrippedAssembly_fi.Length})");
+
+            if (!singleFileBundle)
+            {
+                // original == included assembly
```
Comment (fanyang-mono) [inline src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs:111]: The test is done correctly. However, the comment needs to be updated to `stripped == included assembly`.

**Review by fanyang-mono — APPROVED** [review]
ILStrip test looks good to me.

File: src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs (hunk 1 above)
Comment (fanyang-mono) [reply src/mono/wasi/Wasi.Build.Tests/ILStripTests.cs:15]: reviewed.

**Review by lewing — APPROVED** [review]
Looks ready to merge then we will need to keep an eye on the builds

**Pull request merged**